pkg/volume/scaleio/sio_volume.go

```diff
@@ -0,0 +1,390 @@
+		glog.Error(log("failed to create dir %#v:  %v", dir, err))
+		return err
+	}
+	glog.V(4).Info(log("setup created mount point directory %s", dir))
+
+	diskMounter := &mount.SafeFormatAndMount{
+		Interface: v.plugin.mounter,
+		Runner:    exec.New(),
+	}
+	err = diskMounter.FormatAndMount(devicePath, dir, v.fsType, options)
+
+	if err != nil {
+		glog.Error(log("mount operation failed during setup: %v", err))
+		if err := os.Remove(dir); err != nil && !os.IsNotExist(err) {
+			glog.Error(log("failed to remove dir %s during a failed mount at setup: %v", dir, err))
+			return err
+		}
+		return err
+	}
+
+	glog.V(4).Info(log("successfully setup volume %s attached %s:%s as %s", v.volSpecName, v.volName, devicePath, dir))
+	return nil
+}
+
+func (v *sioVolume) GetAttributes() volume.Attributes {
+	return volume.Attributes{
+		ReadOnly:        v.readOnly,
+		Managed:         !v.readOnly,
+		SupportsSELinux: true,
+	}
+}
+
+// **********************
+// volume.Unmounter Impl
+// *********************
+var _ volume.Unmounter = &sioVolume{}
+
+// TearDownAt unmounts the bind mount
+func (v *sioVolume) TearDown() error {
+	return v.TearDownAt(v.GetPath())
+}
+
+// TearDown unmounts  and remove the volume
+func (v *sioVolume) TearDownAt(dir string) error {
+	v.plugin.volumeMtx.LockKey(v.volSpecName)
+	defer v.plugin.volumeMtx.UnlockKey(v.volSpecName)
+
+	glog.V(4).Info(log("attempting to unmount %s", dir))
+	if err := util.UnmountPath(dir, v.plugin.mounter); err != nil {
+		glog.Error(log("teardown failed while unmounting dir %s: %v ", dir, err))
+		return err
+	}
+	glog.V(4).Info(log("dir %s unmounted successfully", dir))
+
+	// detach/unmap
+	glog.V(4).Info(log("teardown is attempting to detach/unmap volume for %s", v.volSpecName))
+	if err := v.resetSioMgr(); err != nil {
+		glog.Error(log("teardown failed, unable to reset scalio mgr: %v", err))
+	}
+	volName := v.volName
+	if err := v.sioMgr.DetachVolume(volName); err != nil {
+		glog.Error(log("teardown failed detaching volume %s:  %v", volName, err))
+		return err
+	}
+
+	glog.V(4).Infof(log("teardown of volume %v detached successfully", volName))
+	return nil
+}
+
+// ********************
+// volume.Deleter Impl
+// ********************
+var _ volume.Deleter = &sioVolume{}
+
+func (v *sioVolume) Delete() error {
+	glog.V(4).Info(log("deleting pvc %s", v.volSpecName))
+	if err := v.resetSioMgr(); err != nil {
+		glog.Error(log("delete failed, unable to reset scalio mgr: %v", err))
+	}
+	err := v.sioMgr.DeleteVolume(v.volName)
+	if err != nil {
+		glog.Error(log("failed to delete volume %s: %v", v.volName, err))
+		return err
+	}
+
+	glog.V(4).Info(log("successfully deleted pvc %s", v.volSpecName))
+	return nil
+}
+
+// ************************
+// volume.Provisioner Impl
+// ************************
+var _ volume.Provisioner = &sioVolume{}
+
+func (v *sioVolume) Provision() (*api.PersistentVolume, error) {
+	glog.V(4).Info(log("attempting to dynamically provision pvc %v", v.options.PVName))
```

Comment:
what is PVName? It is confusing here because `Provision` is expected to create a PV.